### PR TITLE
drivers/nrf24l01.py Portable driver for Pyboard, ESP8266 and ESP32.

### PR DESCRIPTION
**Pull request by peterhinch**
Please see issue #3366 for description.

### CODE REVIEW

File: drivers/nrf24l01/nrf24l01test.py [hunk 1]
```diff
@@ -1,14 +1,40 @@
-"""Test for nrf24l01 module."""
-
-import struct
+"""Test for nrf24l01 module. Portable between MicroPython targets."""
+try:
+    import struct
+except ImportError:
+    import ustruct as struct
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: Would it work with just `import ustruct as struct` for all boards?  That's a bit cleaner.

File: drivers/nrf24l01/nrf24l01test.py [hunk 2]
```diff
@@ -1,14 +1,40 @@
...
 import utime
 from machine import Pin, SPI
 from nrf24l01 import NRF24L01
+from sys import platform
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: This looks a bit unconventional, I would have used `import sys` and `sys.platform`.

File: drivers/nrf24l01/nrf24l01test.py [hunk 3]
```diff
@@ -1,14 +1,40 @@
...
 
 pipes = (b'\xf0\xf0\xf0\xf0\xe1', b'\xf0\xf0\xf0\xf0\xd2')
 
 def master():
-    nrf = NRF24L01(SPI(2), Pin('Y5'), Pin('Y4'), payload_size=8)
+    csn = Pin(cfg['csn'], mode = Pin.OUT, value = 1)
+    ce = Pin(cfg['ce'], mode = Pin.OUT, value = 0)
+    if cfg['spi'] == -1:
+        spi = SPI(-1, sck=Pin(cfg['sck']), mosi=Pin(cfg['mosi']), miso=Pin(cfg['miso']))
+        nrf = NRF24L01(spi, csn, ce, payload_size=8, baudrate=100000)
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: Does the baudrate need to be specified?  Won't soft SPI just use a maximum of 500kHz if the given value exceeds 500k?

File: drivers/nrf24l01/nrf24l01test.py [hunk 4]
```diff
@@ -60,7 +86,13 @@ def master():
     print('master finished sending; successes=%d, failures=%d' % (num_successes, num_failures))
 
 def slave():
-    nrf = NRF24L01(SPI(2), Pin('Y5'), Pin('Y4'), payload_size=8)
+    csn = Pin(cfg['csn'], mode = Pin.OUT, value = 1)
+    ce = Pin(cfg['ce'], mode = Pin.OUT, value = 0)
+    if cfg['spi'] == -1:
+        spi = SPI(-1, sck=Pin(cfg['sck']), mosi=Pin(cfg['mosi']), miso=Pin(cfg['miso']))
+        nrf = NRF24L01(spi, csn, ce, payload_size=8, baudrate=100000)
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: Same comment as in master() function.

File: drivers/nrf24l01/nrf24l01test.py [hunk 5]
```diff
@@ -60,7 +86,13 @@ def master():
     print('master finished sending; successes=%d, failures=%d' % (num_successes, num_failures))
 
 def slave():
-    nrf = NRF24L01(SPI(2), Pin('Y5'), Pin('Y4'), payload_size=8)
+    csn = Pin(cfg['csn'], mode = Pin.OUT, value = 1)
+    ce = Pin(cfg['ce'], mode = Pin.OUT, value = 0)
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: No spaces around "=" in keyword args.

File: drivers/nrf24l01/nrf24l01test.py [hunk 6]
```diff
@@ -69,7 +101,6 @@ def slave():
     print('NRF24L01 slave mode, waiting for packets... (ctrl-C to stop)')
 
     while True:
-        machine.idle()
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: Did you ever get to the bottom of the issues with this on pyboard?

File: drivers/nrf24l01/nrf24l01test.py [hunk 7]
```diff
@@ -99,9 +132,10 @@ def slave():
...
-print('    SCK on Y6')
-print('    MISO on Y7')
-print('    MOSI on Y8')
+print('    CE on', cfg['ce'])
+print('    CSN on', cfg['csn'])
+print('    SCK on', cfg['sck'])
+print('    MISO on', cfg['miso'])
+print('    MOSI on', cfg['mosi'])
 print('run nrf24l01test.slave() on slave, then nrf24l01test.master() on master')
+
```
Comment (dpgeorge) [inline drivers/nrf24l01/nrf24l01test.py]: Stray blank line?

File: drivers/nrf24l01/nrf24l01test.py (hunk 6 above)
Comment (peterhinch) [reply drivers/nrf24l01/nrf24l01test.py]: No, this has me stumped I'm afraid.

**Comment by peterhinch** [comment]
I'll implement these changes and re-test (in particular for the baudrate).

File: drivers/nrf24l01/nrf24l01test.py (hunk 6 above)
Comment (dpgeorge) [reply drivers/nrf24l01/nrf24l01test.py]: Ok, no worries, it's just a test anyway so not critical to have this line.

**Comment by peterhinch** [comment]
Amended. Re-tested in master and slave roles on each platform.

**Comment by dpgeorge** [comment]
Thanks for updating and retesting.  Merged in ccaa5f5b0bad69f0f0ea81151e34c218fe13ce66 with very minor changes (removing whitespace before ":" in dict constructors, moving "import sys" to first line).

**Pull request closed without merging**